### PR TITLE
[xbmc][cleanup] Clean up ApplicationMessenger

### PR DESCRIPTION
**Pull request by Paxxi**
ApplicationMessenger had become a place to stick all kinds of convenience
methods and depending on just about everything. This PR cleans this up
and returns it back to a messaging class.

Removed the bool parameters and used naming from Win32 API with
SendMsg for blocking and PostMsg for non-blocking messaging.

The message handlers aren't necessarily in the correct place, a bit too many got stuck into CApplication for now. This will need to be taken care of later on.

With this approach it's a tiny bit more cumbersome to send messages but the design means we can use it from any layer without having a GUI dependency, or any real dependency for that matter.

### CODE REVIEW

**Comment by Paxxi** [comment]
@mkortstiege can you fix xcode project for me?

**Comment by mkortstiege** [comment]
Sure, will do tomorrow.

File: xbmc/Application.cpp [hunk 1]
```diff
@@ -1377,17 +1385,17 @@ void CApplication::OnSettingChanged(const CSetting *setting)
     {
       ADDON::AddonPtr addon;
       if (CAddonMgr::Get().GetAddon(((CSettingString*)setting)->GetValue(), addon, ADDON_SKIN))
-      {
+    {
```
Comment (da-anda) [inline xbmc/Application.cpp]: wrong cosmetic change I think

**Comment by Paxxi** [comment]
@mkortstiege I'd like some xcode project work when you have time :)

**Comment by Memphiz** [comment]
@mkortstiege i take it

**Comment by Memphiz** [comment]
@Paxxi cherry-pick https://github.com/Memphiz/xbmc/commit/77bf52894215814e665633000d1486c76624250f - i can't do a PR to your repo (private profile?)

**Comment by Paxxi** [comment]
thanks @Memphiz , it says public so no idea what's wrong, cherrypick works fine so no worries.
Went over everything again and fixed some indentation and whitespace stuff that slipped through.

jenkins build this please

**Comment by Paxxi** [comment]
jenkins build this please

**Comment by Paxxi** [comment]
ok rebasing this is becoming a pain, once this builds cleanly I'll merge it so get reviewing if you so like :)
jenkins build this please

**Comment by Paxxi** [comment]
jenkins build this please

**Comment by Paxxi** [comment]
jenkins build this please

**Comment by Paxxi** [comment]
pray to $deity that jenkins build this please

**Comment by Paxxi** [comment]
getting closer... jenkins build this please

**Comment by Paxxi** [comment]
right, well forgot to actually save the edits...
jenkins build this please for the umpteenth time

**Comment by Paxxi** [comment]
seems jenkins didn't fetch latest changes, let's try adding a dummy
jenkins build this please

**Comment by MilhouseVH** [comment]
Not sure if this is responsible, but there's now a Kodi crash after this PR merged, seems to be PVR related:

http://forum.kodi.tv/showthread.php?tid=231092&pid=2053739#pid2053739

Crash log: http://sprunge.us/LJGE

**Comment by MilhouseVH** [comment]
Crash log from a debug enabled build (Kodi master, OpenELEC master, on RPi), if this helps: http://sprunge.us/ODDd

All I've done is enabled TV with the PVR.Demo client, and this works in build [#0714](http://forum.kodi.tv/showthread.php?tid=231092&pid=2052953#pid2052953). I've then upgraded to build [#0715](http://forum.kodi.tv/showthread.php?tid=231092&pid=2053726#pid2053726) (click link for changelog) which crashes almost immediately after starting up.

Due to extra patches in my OpenELEC builds, serveral line numbers referenced in the stacktrace are a little "off" - this is how the line numbers in the stacktrace map to the github code:

0. `GUIDialog.cpp:214` -> [line 214](https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/GUIDialog.cpp#L214)
1. `GUIWindowManager.cpp:860` -> [line 849](https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/GUIWindowManager.cpp#L849)
2. `ApplicationMessenger.cpp:220` -> [line 220](https://github.com/xbmc/xbmc/blob/master/xbmc/messaging/ApplicationMessenger.cpp#L220)
3. `ApplicationMessenger.cpp:239` -> [line 239](https://github.com/xbmc/xbmc/blob/master/xbmc/messaging/ApplicationMessenger.cpp#L239)
4. `Application.cpp:4526` -> [line 4517](https://github.com/xbmc/xbmc/blob/master/xbmc/Application.cpp#L4517)
5. `XBApplicationEx.cpp:100` -> [line 99](https://github.com/xbmc/xbmc/blob/master/xbmc/XBApplicationEx.cpp#L99)
6. `xbmc.cpp:108` -> [line 108](https://github.com/xbmc/xbmc/blob/master/xbmc/xbmc.cpp#L108)
7. `main.cpp:94` -> [line 79](https://github.com/xbmc/xbmc/blob/master/xbmc/main/main.cpp#L79)

File: xbmc/interfaces/legacy/Player.cpp [hunk 2]
```diff
@@ -98,12 +100,13 @@ namespace XBMCAddon
...
-          CApplicationMessenger::Get().PlayFile((const CFileItem)(*(listitem->item)), false);
+          CApplicationMessenger::Get().PostMsg(TMSG_MEDIA_PLAY, 0, 0, static_cast<void*>(new CFileItem(*listitem->item)));
         }
         else
         {
-          CFileItem nextitem(item, false);
-          CApplicationMessenger::Get().MediaPlay(nextitem.GetPath());
+          CFileItemList *l = new CFileItemList; //don't delete,
+          l->Add(std::make_shared<CFileItem>(item, false));
+          CApplicationMessenger::Get().PostMsg(TMSG_MEDIA_PLAY, -1, -1, static_cast<void*>(l));
```
Comment (xhaggi) [inline xbmc/interfaces/legacy/Player.cpp]: should we think about a helper class for all of those convenience methods? less code duplications and easier to use.

**Comment by uNiversaI** [comment]
@paxxi http://trac.kodi.tv/ticket/16129 all reported there strongly indicates caused by this PR.

File: xbmc/AppParamParser.cpp [hunk 3]
```diff
@@ -35,6 +35,8 @@
 #endif
 #include <stdlib.h>
 
+using namespace KODI::MESSAGING;
```
Comment (FernetMenta) [inline xbmc/AppParamParser.cpp]: this is misleading. AppMessanger is all but not Kodi's messaging system

Comment (Paxxi) [inline xbmc/AppParamParser.cpp]: After the rework it only does message passing so it felt suitable

Comment (FernetMenta) [inline xbmc/AppParamParser.cpp]: AppMessenger is used if you want something executed by the main thread. AE, VDPAU, VAAPI, dvdplayer have messaging as well but are not ties to this namespace. This is mainthread only and should be flagged as such.

Comment (Paxxi) [inline xbmc/AppParamParser.cpp]: A bit late to start arguing over namespace naming now, pr was open for a month so there was plenty of time for that

Comment (FernetMenta) [inline xbmc/AppParamParser.cpp]: > was plenty of time for that

don't forget that we are short final for 15.0. So not plenty of time for someone who cares about this release. 

> A bit late to start arguing over namespace naming 

it is never too late to correct something that went wrong.

**Comment by AchimTuran** [comment]
@Paxxi 
@FernetMenta and @fritsch What do you think about messages in Kodi? I think you both are very familiar with AE's system.

I know this is already closed, but I have some question about the new CApplicationMessenger. Don't misunderstand me I like the idea to send messages around in Kodi and I also like the simple API.

But I don't understand why we have now two message systems. AE has already a well profed and feature rich system. Why do we not use this system for all modules in Kodi?
See AE's protocoll and message class [here](https://github.com/xbmc/xbmc/blob/master/xbmc/utils/ActorProtocol.h)

I also miss a response from CApplicationMessenger, when a message fails or we get an error from the system. Today the system crashed, because I tried to send a message in CApplication::Create(). The problem was that no receivers were registered. See my last [commits](https://github.com/AlwinEsch/kodi/commits/audio-dsp-addon-handling)

**Comment by Paxxi** [comment]
@AchimTuran The crash is obviously a bug and I'll investigate it. Sending messages that requires responses are currently handled via the void* parameter.

I was thinking of adding the ability to specify a thread to run on to make it more suitable for non ui messages as well but that's still in the planning stages

**Comment by AchimTuran** [comment]
@Paxxi 
If you want I can create a new PR for it and then we can improve it.

**Comment by uNiversaI** [comment]
@Paxxi Sorry to pester you, the web interface when selecting the remote control, you cant go back from that or turn off remote pressing same button like before.

Also the PLAY works now but sometimes takes 2 or 3 clicks to get the media playing.

**Comment by Paxxi** [comment]
@un1versal gimme debug log and I'll look into it.

**Comment by uNiversaI** [comment]
@paxxi http://pastebin.com/757yZQZK extra json logging enabled played a track via iwebinterface entered remote control on webineface  pressed left/right then pressed the remote button to go off web interface > no dice.

**Comment by Paxxi** [comment]
@uNiversaI thanks, also it's not pestering if I caused bugs then I want to know everything about it :)

@AchimTuran that will come later on, my next goal is to now use this to get rid of all dialog creation in non ui code.

**Comment by Paxxi** [comment]
those changes you mention for thte remote, that's all for the web remote? those don't have anything to do with this pr.

**Comment by uNiversaI** [comment]
Its part of the web interface which use to work and this PR broke.

**Pull request merged**